[UXIT-1307] Digest Page Scaffold

barbaraperic opened the pull request:
## 📝 Description

This PR includes a new `digest` and `digest/[slug]` pages. 

- **Type:** New feature

## 🛠️ Key Changes

- `digest` and `digest/[slug]` pages and related tests and config files
- new cms config schema 
- updated navigation 

## 📌 To-Do Before Merging

- [ ] Remove dummy digest article
- [ ] Maybe also delete `app/_data/cmsConfigSchema.json`?

## 🧪 How to Test
- **Steps to Test:**
  1. Navigate to the content manager and create a new digest article.
  2. Verify that the new digest article appears correctly on the main digest page.
  3. Click on the digest article to open its dedicated digest/[slug] page.
  4. Confirm that the layout, styling, and content display as expected, matching the Figma designs.

## 📸 Screenshots

![screencapture-localhost-3000-digest-2024-08-27-10_30_09](https://github.com/user-attachments/assets/b1909343-1997-4274-ac7c-5cf90a82b382)
![screencapture-localhost-3000-digest-letter-from-the-guest-editor-2024-08-27-10_30_56](https://github.com/user-attachments/assets/0ef4fab3-a8b3-42ff-a5a4-17a849dbdc54)

mirhamasala commented on src/app/_data/cmsConfigSchema.json:
Needs to be deleted.

mirhamasala commented on src/content/digest/letter-from-the-guest-editor.md:
```diff
@@ -1,47 +1,41 @@
-image:
-  src: /assets/images/01-lockheed-p-1080.png
+article-number: 1
+publication-date: 2024-08-27T10:15:00.000Z
```
What is this for? Where in the design it used?

mirhamasala replied:
From conversations I understood with Gary and Filipa, we just want to have the issue date.

mirhamasala reviewed (changes requested):
@barbaraperic I pushed a few CRs. Please have a look before asking @CharlyMartin for a review. 🙏🏼

We should move the blog post and digest article header components within their respective pages components folder once all done.

barbaraperic commented:
> @barbaraperic I pushed a few CRs. Please have a look before asking @CharlyMartin for a review. 🙏🏼
> 
> ~We should move the blog post and digest article header components within their respective pages components folder once all done.

Thank you @mirhamasala 🙏

Reviewed and added:
- refactored `Card` to take a list of tags [reflecting the updated design ](https://filecoinfoundation.slack.com/archives/C06MTEQ3SP6/p1724849356701029)
- order cards by issue number
- added `extractDescriptionFromContent` function to prevent the below of happening - case when add a title before the paragraph
<img width="678" alt="Screenshot 2024-08-28 at 13 57 30" src="https://github.com/user-attachments/assets/96b593a6-7be9-41a1-9b07-efdc75a35cd7">

mirhamasala commented on src/app/_components/ArticleHeader.tsx:
```diff
@@ -0,0 +1,46 @@
+          priority
+          quality={100}
+          src={image?.src || ''}
+          alt={image?.alt || 'Article Header Image'}
```
```suggestion
          alt={image?.alt || ''}
```

CharlyMartin commented on src/app/_components/DigestArticleHeader.tsx:
```diff
@@ -0,0 +1,52 @@
+  )
+}
+
+export function formatAuthors(authors: DigestArticleData['authors']) {
+  return authors
+    .map((author, index) => {
+      const isLastAuthor = index === authors.length - 1
+      const isSecondToLastAuthor = index === authors.length - 2
+
+      let separator = ''
+      if (isSecondToLastAuthor && authors.length > 1) {
+        separator = ' & '
+      } else if (!isLastAuthor) {
+        separator = ', '
+      }
+
+      return `${author.firstName} ${author.lastName}${separator}`
+    })
+    .join('')
+}
```
I find this function a little bit confusing to read. I came up with this version, what do you think?

```suggestion
export function formatAuthors(authors: DigestArticleData['authors']) {
  const COMMA = ', '
  const AMPERSAND = ' & '

  if (authors.length === 0) {
    return ''
  }

  if (authors.length === 1) {
    const [onlyAuthor] = authors
    return buildAuthorName(onlyAuthor)
  }

  if (authors.length === 2) {
    return authors.map(buildAuthorName).join(AMPERSAND)
  }

  const lastAuthor = authors.pop()!
  return (
    authors.map(buildAuthorName).join(COMMA) +
    AMPERSAND +
    buildAuthorName(lastAuthor)
  )
}

function buildAuthorName(author: DigestArticleData['authors'][number]) {
  return `${author.firstName} ${author.lastName}`
}
```

mirhamasala replied:
Personally, I prefer the initial function.

CharlyMartin commented on src/app/_utils/createTagsProp.ts:
```diff
@@ -0,0 +1,11 @@
+import type { TagProps } from '@/components/TagLabel'
+
+export function createTagsProp({
+  label,
+  borderColor,
+}: {
+  label: TagProps['children']
+  borderColor?: TagProps['borderColor']
+}): Array<TagProps> {
+  return [{ children: label, borderColor }]
+}
```
```suggestion
type CreateTagsPropParams = {
  label: TagProps['children']
  borderColor?: TagProps['borderColor']
}

export function createTagsProp({ label, borderColor }: CreateTagsPropParams) {
  return [{ children: label, borderColor }]
}
```

CharlyMartin commented on src/app/_utils/sortDigestArticles.ts:
```diff
@@ -0,0 +1,11 @@
+import { type DigestArticleData } from '@/types/digestTypes'
+
+export function sortDigestArticles(
+  a: DigestArticleData,
+  b: DigestArticleData,
+): number {
+  if (Number(a.issueNumber) !== Number(b.issueNumber)) {
+    return Number(a.issueNumber) - Number(b.issueNumber)
+  }
+  return a.articleNumber - b.articleNumber
+}
```
```suggestion
export function sortDigestArticles(a: DigestArticleData, b: DigestArticleData) {
  if (Number(a.issueNumber) !== Number(b.issueNumber)) {
    return Number(a.issueNumber) - Number(b.issueNumber)
  }

  return a.articleNumber - b.articleNumber
}
```

Also, maybe the function name could be more descriptive? How are we sorting the digest articles? Ascending or descending?

barbaraperic replied:
how's `sortDigestArticlesAscendingByIssueAndNumber`?

CharlyMartin commented on src/app/_utils/extractDescriptionFromContent.ts:
```diff
@@ -0,0 +1,13 @@
+export function extractDescriptionFromContent(content: string): string {
+  const lines = content.split('\n')
+  const startIndex = lines[0].startsWith('#') ? 1 : 0
+
+  return (
+    lines
+      .slice(startIndex)
+      .join(' ')
+      .replace(/\s+/g, ' ')
+      .trim()
+      .split('.')[0] + '.'
+  )
+}
```
I'm not sure what that function is doing, but I still see `#` characters in the Card preview on my end. Here's the content I used for reference:

```
# Tum edidit somnia pennas

## Non et cogeret adiit de

Lorem markdownum ignis vulnere sede robora e placare figura. Quales per Pheneon
emensas flumina!

## Fata nec indignantia

Dimittite unus considere fugit erat cruor Corythus, pondera parmam cedendo omne
ardent. Mitior cernis donec mane, fata est mentisque oscula relatu, cantat
_numina fallite_, et vere, adyto.

## Favorem geri hunc videre inmansuetumque sua tarda

Venit exit est dextro alta modo illum artesque Minyae Aeaciden procul. Corpora
iura rupem deos sano Caesareo, veteris utque. Deceant cadunt his nulla deos: das
gemitu, corpus abstractus, vidisse?
```

mirhamasala replied:
Let's discuss this function tomorrow together..

CharlyMartin commented on src/app/digest/page.tsx:
```diff
@@ -0,0 +1,105 @@
+                title={title}
+                description={extractDescriptionFromContent(content)}
+                textIsClamped={true}
+                tags={[
+                  ...createTagsProp({
+                    label: `Issue ${issueNumber}`,
+                    borderColor: 'brand-100',
+                  }),
+                  ...createTagsProp({
+                    label: `Article ${articleNumber}`,
+                  }),
+                ]}
```
Could `createTagsProp` take an object OR an array of objects as argument? It would make more semantic sense to me, but open to suggestions :)

mirhamasala replied:
I agree that this needs to be reworked ... we need to look into a better solution.

CharlyMartin commented on src/app/digest/[slug]/page.tsx:
```diff
@@ -0,0 +1,50 @@
+          authors={authors}
+        />
+        {content && <MarkdownContent>{content}</MarkdownContent>}
+      </div>
```
```suggestion
        <MarkdownContent>{content}</MarkdownContent>
```

`content` is always a string right? So the `&&` check is unnecessary?

CharlyMartin commented on src/app/digest/page.tsx:
```diff
@@ -0,0 +1,105 @@
+      />
+
+      <PageSection
+        kicker="Issue 1 - Sep 2024"
```
```suggestion
        kicker="Issue 1 - Sept 2024"
```

barbaraperic replied:
Actually in the design is `Sep` (we are also using `Sep` for the events)

CharlyMartin commented:
@barbaraperic I added my review with a bunch of suggestions and comments.

About the test steps:
I created a dummy digest article and the Zod validations failed because I didn’t add an author in the CMS.
Is it possible to make that field required in the CMS?

![CleanShot 2024-08-29 at 15 25 31@2x](https://github.com/user-attachments/assets/65a10d99-45ac-4204-ad97-90d7dab3ac82)

Once I added the author manually in the Markdown file, it worked fine 👏

mirhamasala commented on src/app/_components/Card.tsx:
```diff
@@ -131,9 +131,13 @@ export function Card({
+        {tags && tags.length > 0 && (
+          <span className="mb-4 flex gap-2">
+            {tags.map((tag) => (
+              <TagLabel key={`${tag.children}`} borderColor={tag.borderColor}>
```
I agree with Charly - It's confusing to me.
